singer/src/main/java/com/pinterest/singer/writer/kafka/CommittableKafkaWriter.java

```diff
@@ -290,6 +249,56 @@ public void endCommit(int numLogMessages) throws LogStreamWriterException {
     }
   }
 
+  private void captureAndLogAuditEvents(KafkaWritingTaskResult result) {
+    int bucketIndex = result.getPartition();
+    // when result.success is true, the number of recordMetadata SHOULD be the same
+    // as the number of ProducerRecord. The size mismatch should never happen.
+    // Adding this if-check is just an additional verification to make sure the size
+    // match and the audit events sent out is indeed corresponding to those log messages
+    // that are audited.
+    List<Future<RecordMetadata>> recordMetadataList = commitableBuckets.get(bucketIndex)
+        .getRecordMetadataList();
+    if (bucketIndex >= 0
+        && result.getRecordMetadataList().size() != recordMetadataList.size()) {
+      // this should never happen!
+      LOG.warn(
+          "Number of ProducerRecord does not match the number of RecordMetadata, "
+              + "LogName:{}, Topic:{}, BucketIndex:{}, result_size:{}, bucket_size:{}",
+          logName, topic, bucketIndex, result.getRecordMetadataList().size(),
+          recordMetadataList.size());
+      OpenTsdbMetricConverter.incr(SingerMetrics.AUDIT_HEADERS_METADATA_COUNT_MISMATCH, 1,
+          "topic=" + topic, "host=" + HOSTNAME, "logStreamName=" + logName,
+          "partition=" + bucketIndex);
+    } else {
+      // regular code execution path
+      enqueueLoggingAuditEvents(result, commitableMapOfHeadersMap.get(bucketIndex));
+      OpenTsdbMetricConverter.incr(SingerMetrics.AUDIT_HEADERS_METADATA_COUNT_MATCH, 1,
+          "host=" + HOSTNAME, "logStreamName=" + logName);
+    }
+  }
+
+  private void updateWriteFailureMetrics(int numLogMessages) {
+    OpenTsdbMetricConverter.incr("singer.writer.producer_reset", 1, "topic=" + topic,
+        "host=" + HOSTNAME);
+    OpenTsdbMetricConverter.incr("singer.writer.num_kafka_messages_delivery_failure",
+        numLogMessages, "topic=" + topic, "host=" + HOSTNAME, "logname=" + logName);
```

Comment:
Should these metric names also be defined in `SingerMetrics` class as constants?

Reply:
Yes, but the metric name should be changed to num_messages_delivery. Once we are able to refactor the names we should be able move these to SingerMetrics